lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,114 @@
+//===-- lib/semantics/check-select-stmt.cpp -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "check-select-stmt.h"
+#include "flang/common/Fortran.h"
+#include "flang/common/idioms.h"
+#include "flang/parser/message.h"
+#include "flang/parser/tools.h"
+#include "flang/semantics/tools.h"
+#include <list>
+#include <optional>
+#include <set>
+#include <tuple>
+#include <variant>
+
+namespace Fortran::semantics {
+
+void SelectStmtChecker::Leave(
+    const parser::SelectRankConstruct &selectRankConstruct) {
+  const auto &selectRankStmt{
+      std::get<parser::Statement<parser::SelectRankStmt>>(
+          selectRankConstruct.t)};
+  const auto &selectRankStmtSel{
+      std::get<parser::Selector>(selectRankStmt.statement.t)};
+
+  // R1149 select-rank-stmt checks
+  const Symbol *saveSelSymbol{nullptr};
+  if (const auto resolvedSel{ResolveSelector(selectRankStmtSel)}) {
+    if (const Symbol * sel{evaluate::UnwrapWholeSymbolDataRef(*resolvedSel)}) {
+      if (!IsAssumedRankArray(*sel)) {
+        context_.Say(parser::FindSourceLocation(selectRankStmtSel),
+            "Selector is not an assumed-rank array variable"_err_en_US);
+      } else {
+        saveSelSymbol = sel;
+      }
+    } else {
+      context_.Say(parser::FindSourceLocation(selectRankStmtSel),
+          "Selector is not an assumed-rank array variable"_err_en_US);
+    }
+  }
+
+  // R1150 select-rank-case-stmt checks
+  auto &rankCaseList{std::get<std::list<parser::SelectRankConstruct::RankCase>>(
+      selectRankConstruct.t)};
+  bool defaultRankFound{false};
+  bool starRankFound{false};
+  std::set<std::int64_t> matches;
```

Comment:
Since this set only has to hold the values in [0..maxRank], a bitset would work and might be more clear.  But even better would be a simple array of optional source locations (`std::optional<CharBlock> caseForRank[maxRank + 1];`), so that your error message about duplicate rank cases could point to the preceding duplicate case when it exists.